backend/src/map_model.rs

```diff
@@ -1188,12 +1188,31 @@ impl DiagonalFilter {
             .map(|offset| intersection.roads[(offset + split_offset) % intersection.roads.len()])
             .collect();
 
-        let road_1 = map_model.get_r(group_a[0]);
-        let road_2 = map_model.get_r(group_a[1]);
+        let angle = {
+            let road_a = if is_rotated {
+                map_model.get_r(group_a[0])
+            } else {
+                map_model.get_r(group_a[1])
+            };
+            let road_b0 = map_model.get_r(group_b[0]);
+            let road_b1 = map_model.get_r(group_b[1]);
+
+            let bearing_a0 = bearing_from_endpoint(intersection.point, &road_a.linestring);
+            let bearing_b0 = bearing_from_endpoint(intersection.point, &road_b0.linestring);
+            let bearing_b1 = bearing_from_endpoint(intersection.point, &road_b1.linestring);
+
+            let angle_b0 = angle_between_bearings(bearing_a0, bearing_b0);
+            let angle_b1 = angle_between_bearings(bearing_a0, bearing_b1);
 
-        let bearing_1 = bearing_from_endpoint(intersection.point, &road_1.linestring);
-        let bearing_2 = bearing_from_endpoint(intersection.point, &road_2.linestring);
-        let angle = diagonal_bearing(bearing_1, bearing_2) as f32;
+            // Split the acute angle, since the obtuse angle has more tolerance
```

Comment:
This is the key - rather than splitting between arbitrary roads, we make sure our split is relative to the most acute angle, which has the tightest tolerance.